Win32-network/test/Test/Async.hs

```diff
@@ -0,0 +1,286 @@
+              tid' <- forkIO (void $ Win32.Async.readHandle hpipe' 1)
+              threadDelay 100
+              killThread tid
+              killThread tid'
+
+
+test_interruptible_writeHandle :: IO ()
+test_interruptible_writeHandle = Win32.Async.withIOManager $ \iocp -> do
+    let bs = BSC.pack $ replicate 100 'a'
+    v <- newEmptyMVar
+    syncVar <- newEmptyMVar
+
+    bracket
+      ((,) <$> createNamedPipe pipeName
+                               (pIPE_ACCESS_DUPLEX .|. fILE_FLAG_OVERLAPPED)
+                               (pIPE_TYPE_BYTE .|. pIPE_READMODE_BYTE)
+                               pIPE_UNLIMITED_INSTANCES
+                               1
+                               1
+                               0
+                               Nothing
+           <*> createFile pipeName
+                          gENERIC_WRITE
+                          fILE_SHARE_NONE
+                          Nothing
+                          oPEN_EXISTING
+                          fILE_FLAG_OVERLAPPED
+                          Nothing)
+      (foldMap closeHandle)
+      $ \(r, w) -> do
+        _ <- Win32.Async.associateWithIOCompletionPort r iocp
+        _ <- Win32.Async.associateWithIOCompletionPort w iocp
+
+        tid <- forkIOWithUnmask $ \unmask ->
+          void $ do
+            putMVar syncVar ()
+            unmask (Win32.Async.writeHandle w bs)
+          `catch` \(e :: AsyncException) -> putMVar v e >> throwIO e
+
+        -- wait for 'writeHandle'
+        takeMVar syncVar
+        threadDelay 100
+        killThread tid
+
+        e <- takeMVar v
+        assertBool "test_interruptible_writeHandle" (ThreadKilled == e)
+
+-- | Run a server and client which both simultanously read and write from a
+-- handle.
+--
+prop_async_read_and_writes :: LargeNonEmptyBS
+                           -> LargeNonEmptyBS
+                           -> Property
+prop_async_read_and_writes (LargeNonEmptyBS bsIn bufSizeIn) (LargeNonEmptyBS bsOut bufSizeOut) =
+    ioProperty $ Win32.Async.withIOManager $ \iocp -> do
+      threadDelay 100
+
+      syncVarStart <- newEmptyMVar
+      syncVarEnd   <- newEmptyMVar
+      clientVar <- newEmptyMVar
+      serverVar <- newEmptyMVar
+
+      -- fork a server
+      _ <- forkIO $
+        bracket
+            (createNamedPipe pipeName
+                             (pIPE_ACCESS_DUPLEX .|. fILE_FLAG_OVERLAPPED)
+                             (pIPE_TYPE_BYTE .|. pIPE_READMODE_BYTE)
+                             pIPE_UNLIMITED_INSTANCES
+                             (fromIntegral bufSizeIn)
+                             (fromIntegral bufSizeOut)
+                             0
+                             Nothing)
+            closeHandle
+            $ \h -> do
+              -- associate 'h' with  I/O completion 'port'
+              _ <- Win32.Async.associateWithIOCompletionPort h iocp
+              putMVar syncVarStart ()
+              connectNamedPipe h Nothing
```

Comment:
Here's where we're using `connectNamedPipe` with a pipe in overlapped mode, and we'll be doing the same thing for real, so we'll have to make this properly async, as per the comment on the definition.